### PR TITLE
DEPR: Deprecate use of strings denoting units with 'M', 'Y' or 'y' in pd.to_timedelta (36666)

### LINKED ISSUE #36666: DEPR deprecate pd.to_timedelta('1 M') and pd.to_timedelta('1 Y')
We have that `pd.to_timedelta(1, unit='M')` and `pd.to_timedelta(1, unit='Y')` are deprecated, and so so should these (xref https://github.com/pandas-dev/pandas/pull/34979#issuecomment-699494617 ).

Once deprecated, the note added in #34979 should be removed

### PR DESCRIPTION
**Pull request by avinashpancham**
- [x] closes #36666
- [x] tests added / passed
- [x] passes `black pandas`
- [x] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [x] whatsnew entry

### CODE REVIEW

**Comment by avinashpancham** [comment]
Do we only want to deprecate variations with a space (`pd.to_timedelta("1 Y")`) or also without a space  (`pd.to_timedelta("1Y")`)

File: pandas/core/tools/timedeltas.py [hunk 1]
```diff
@@ -25,10 +26,13 @@ def to_timedelta(arg, unit=None, errors="raise"):
...
-        365.2425 days or 365 days 5 hours 49 minutes 12 seconds.
+        The data to be converted to timedelta.
+
+        .. deprecated:: 1.2
+            Strings denoting units with 'M', 'Y', 'm' or 'y' do not represent
+            unambiguous timedelta values durations and will removed in a future
+            version
+
     unit : str, optional
         Denotes the unit of the arg for numeric `arg`. Defaults to ``"ns"``.
```
Comment (avinashpancham) [inline pandas/core/tools/timedeltas.py:37]: Off-topic:  parameter unit defaults to ``"ns"`` according to docstring, but the python code defaults it to `None`. Can I change it here or should there be a separate PR for this?

Comment (mroeschke) [reply pandas/core/tools/timedeltas.py:37]: We use this pattern commonly in pandas. Fine to leave as is.

File: pandas/core/tools/timedeltas.py [hunk 2]
```diff
@@ -118,8 +122,18 @@ def to_timedelta(arg, unit=None, errors="raise"):
...
         )
 
-    if isinstance(arg, str) and unit is not None:
-        raise ValueError("unit must not be specified if the input is/contains a str")
+    if isinstance(arg, str):
+        if unit is not None:
+            raise ValueError(
+                "unit must not be specified if the input is/contains a str"
+            )
+        elif arg.upper().endswith(" M") or arg.upper().endswith(" Y"):
```
Comment (mroeschke) [inline pandas/core/tools/timedeltas.py:130]: As mentioned in your other commented, we also want to catch arguments without the space as well e.g. `"1Y"`. 

Also we need a test for this warning

Comment (avinashpancham) [reply pandas/core/tools/timedeltas.py]: Done

**Comment by avinashpancham** [comment]
Ah I see some tests are failing. I will update the code.

**Comment by avinashpancham** [comment]
All tests related to `pd.to_timedelta` are passing. The failed tests originate from another issue on master.

**Comment by avinashpancham** [comment]
Updated my code and merged master. Everything passes now.

File: pandas/tests/tools/test_to_timedelta.py [hunk 3]
```diff
@@ -121,6 +121,25 @@ def test_to_timedelta_invalid(self):
...
+            ("1m", FutureWarning),
+            ("1 m", FutureWarning),
+            ("1y", FutureWarning),
+            ("1 y", FutureWarning),
+            ("1 day", None),
+            ("2day", None),
+        ],
+    )
+    def test_unambiguous_timedelta_values(self, val, warning):
+        with tm.assert_produces_warning(warning):
```
Comment (mroeschke) [inline pandas/tests/tools/test_to_timedelta.py:140]: Could you but the github issue number here as a comment?

Comment (avinashpancham) [reply pandas/tests/tools/test_to_timedelta.py:142]: Done. CI fails due to other tests on master though.

File: pandas/core/tools/timedeltas.py [hunk 4]
```diff
@@ -118,8 +123,18 @@ def to_timedelta(arg, unit=None, errors="raise"):
             "arg must be a string, timedelta, list, tuple, 1-d array, or Series"
         )
 
-    if isinstance(arg, str) and unit is not None:
-        raise ValueError("unit must not be specified if the input is/contains a str")
+    if isinstance(arg, str):
+        if unit is not None:
+            raise ValueError(
+                "unit must not be specified if the input is/contains a str"
+            )
```
Comment (jreback) [inline pandas/core/tools/timedeltas.py:130]: hmm we *already* have this check on L103, so something must be going wrong here.

File: pandas/tests/scalar/timedelta/test_timedelta.py [hunk 5]
```diff
@@ -258,7 +258,9 @@ def test_unit_parser(self, units, np_unit, wrapper):
             if unit == "M":
                 expected = Timedelta(np.timedelta64(2, "m").astype("timedelta64[ns]"))
 
-            result = to_timedelta(f"2{unit}")
+            warning = None if unit != "m" else FutureWarning
```
Comment (jreback) [inline pandas/tests/scalar/timedelta/test_timedelta.py:261]: change this to parameterize this (e.g. elminate the for loop and parameterize on units)

then add a parameter whether should test for FutureWarning or not (could be bool is ok).

then in the body you can easily check this for both `to_timedelta` and `Timedelta`

**Review by jreback — CHANGES_REQUESTED** [review]

File: pandas/core/tools/timedeltas.py (hunk 4 above)
Comment (avinashpancham) [reply pandas/core/tools/timedeltas.py:130]: if you do: `pd.to_timedelta("1 Y")` then unit is not specified and by default None. So then you never reach this code.

File: pandas/tests/scalar/timedelta/test_timedelta.py (hunk 5 above)
Comment (avinashpancham) [reply pandas/tests/scalar/timedelta/test_timedelta.py:261]: We have approx 50 values that have no warning and just one ("m") that has a warning. IMO a parameter for this would be overkill. But if we want this then I can change it.

Comment (avinashpancham) [reply pandas/tests/scalar/timedelta/test_timedelta.py]: Updated the test. IMO `test_unit_parser` could refactored even further: split in 2 tests: `test_unit_parser_array` and `test_unit_parser_scalar` and apply a parameter on the function that converts its to timedelta. Do we want to do this?

**Comment by avinashpancham** [comment]
I found out that the following was also possible: `pd.to_timedelta(["1Y"])` or `pd.to_timedelta(pd.Series(["1Y"]))`. So I updated `to_timedelta` to also give depr warnings in this situation. I'm not completely happy with this added code, so any feedback is appreciated

File: pandas/core/tools/timedeltas.py [hunk 6]
```diff
@@ -101,6 +110,16 @@ def to_timedelta(arg, unit=None, errors="raise"):
             "represent unambiguous timedelta values durations."
         )
 
+    if isinstance(arg, (list, tuple, ABCSeries, ABCIndexClass, np.ndarray)):
```
Comment (jreback) [inline pandas/core/tools/timedeltas.py:113]: the warning shouldn't be here

File: pandas/_libs/tslibs/timedeltas.pyx [hunk 7]
```diff
@@ -1467,6 +1468,16 @@ cdef _broadcast_floordiv_td64(
         return res
 
 
+def check_unambiguous_timedelta_values(object[:] values):
```
Comment (jreback) [inline pandas/_libs/tslibs/timedeltas.pyx:1471]: rather than this you can do it in the Timedelta string converter

File: pandas/core/tools/timedeltas.py [hunk 8]
```diff
@@ -118,8 +137,18 @@ def to_timedelta(arg, unit=None, errors="raise"):
...
 
-    if isinstance(arg, str) and unit is not None:
-        raise ValueError("unit must not be specified if the input is/contains a str")
+    if isinstance(arg, str):
+        if unit is not None:
+            raise ValueError(
+                "unit must not be specified if the input is/contains a str"
+            )
+        elif re.search(r"^\d+\s?[M|Y|m|y]$", arg):
+            warnings.warn(
```
Comment (jreback) [inline pandas/core/tools/timedeltas.py:146]: i wouldn't do this here at all

**Review by jreback — CHANGES_REQUESTED** [review]
you need to do the warning inside the string to timedelta conversion code in timedelta.pyx itself;

**Comment by avinashpancham** [comment]
Thanks @jreback , moved the warning to timedelta.pyx and the code is much cleaner now.

File: doc/source/whatsnew/v0.15.0.rst [hunk 9]
```diff
@@ -1030,6 +1030,7 @@ Other:
   If ``Period`` freq is ``D``, ``H``, ``T``, ``S``, ``L``, ``U``, ``N``, ``Timedelta``-like can be added if the result can have same freq. Otherwise, only the same ``offsets`` can be added.
 
   .. ipython:: python
+    :okwarning:
```
Comment (jreback) [inline doc/source/whatsnew/v0.15.0.rst:1033]: why is this needed?

File: pandas/_libs/tslibs/timedeltas.pyx [hunk 10]
```diff
@@ -297,6 +299,14 @@ cdef inline int64_t parse_timedelta_string(str ts) except? -1:
     if len(ts) == 0 or ts in nat_strings:
         return NPY_NAT
 
+    if re.search(r"^\d+\s?[M|Y|m|y]$", ts):
+        warnings.warn(
```
Comment (jreback) [inline pandas/_libs/tslibs/timedeltas.pyx:303]: you will instead want to do this in timespec_from_spec and not use a regex (they are slow compared to what is going on).

File: pandas/tests/resample/test_datetime_index.py [hunk 11]
```diff
@@ -752,7 +752,8 @@ def test_resample_origin():
     resampled = ts.resample("5min", origin=offset_timestamp).mean()
     tm.assert_index_equal(resampled.index, exp_rng)
 
-    resampled = ts.resample("5min", origin="epoch", offset="2m").mean()
+    with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
```
Comment (jreback) [inline pandas/tests/resample/test_datetime_index.py:755]: why are you catching this? why does this have a warning?

File: pandas/tests/scalar/timedelta/test_constructors.py [hunk 12]
```diff
@@ -40,8 +41,9 @@ def test_construction():
     assert Timedelta("1 hr") == timedelta(hours=1)
     assert Timedelta("1 hours") == timedelta(hours=1)
     assert Timedelta("-1 hours") == -timedelta(hours=1)
-    assert Timedelta("1 m") == timedelta(minutes=1)
-    assert Timedelta("1.5 m") == timedelta(seconds=90)
+    with tm.assert_produces_warning(FutureWarning, check_stacklevel=False):
+        assert Timedelta("1 m") == timedelta(minutes=1)
```
Comment (jreback) [inline pandas/tests/scalar/timedelta/test_constructors.py:45]: there is nothing wrong here, I think you are catching m for some reason

**Review by jreback — CHANGES_REQUESTED** [review]

File: doc/source/whatsnew/v0.15.0.rst (hunk 9 above)
Comment (avinashpancham) [reply doc/source/whatsnew/v0.15.0.rst:1033]: That ipython block contains the code`pd.Timdelta('120m')` that due to the current changes gives a warning because of the use of m. As a result the CI check in web and docs failed and it was suggested to add the :okwarning: line

File: pandas/_libs/tslibs/timedeltas.pyx (hunk 10 above)
Comment (avinashpancham) [reply pandas/_libs/tslibs/timedeltas.pyx]: Thanks for the tip, moved it.

File: pandas/tests/resample/test_datetime_index.py (hunk 11 above)
Comment (avinashpancham) [reply pandas/tests/resample/test_datetime_index.py]: I thought we shouldnt change tests, hence the assert. Now I removed the assert and changed m to min.

File: pandas/tests/scalar/timedelta/test_constructors.py (hunk 12 above)
Comment (avinashpancham) [reply pandas/tests/scalar/timedelta/test_constructors.py]: I thought we shouldnt change tests, hence the assert. Now I removed the assert and changed m to min.

File: doc/source/whatsnew/v0.15.0.rst (hunk 9 above)
Comment (avinashpancham) [reply doc/source/whatsnew/v0.15.0.rst]: Today I found #6856 where it was proposed to change the ipython directives with code-blocks to prevent this issue. I have also applied that logic here

**Comment by avinashpancham** [comment]
@jreback implemented all suggestions, could you have a 2nd look? CI is still failing after merging master, but that is related to other issues than introduced in this PR.

**Comment by avinashpancham** [comment]
CI failed due to FATAL ERROR: Committing semi space failed. Allocation failed - process out of memory. Can someone please restart it? Thanks!

**Comment by jreback** [comment]
why is 'm' (lowercase) m deprecated? this is minutes as in not ambiguous.  (agree on Y, y, M)

**Comment by avinashpancham** [comment]
My bad, I thought we wanted to deprecate all variations of m (lowercase and uppercase). Will revert it then.

**Comment by jreback** [comment]
> My bad, I thought we wanted to deprecate all variations of m (lowercase and uppercase). Will revert it then.

kk great, otherwise lgtm. ping  on green.

**Comment by avinashpancham** [comment]
@jreback reverted the depr of lowercase m, merged master and CI on green.

**Review by jreback — APPROVED** [review]

**Comment by jreback** [comment]
thanks @avinashpancham very nice!
